cli/src/commands/git.rs

```diff
@@ -189,6 +189,28 @@ pub struct GitCloneArgs {
 /// `remote_branches(remote=<remote>)..@`. Use `--branch` to push specific
 /// branches. Use `--all` to push all branches. Use `--change` to generate
 /// branch names based on the change IDs of specific commits.
+///
+/// Before the command actually moves or creates a remote branch, it makes
+/// several safety checks.
+///
+/// 1. `jj` will contact the remote and check that the actual state of the
+///    remote branch matches `jj`'s record of its last known position. If they
+///    do not match, `jj` will refuse to push the branch. In this case, you need
+///    to `jj git fetch --remote <remote name>`. This may create some branch
+///    conflicts, which you'll need to resolve before trying `jj git push`
+///    again.
```

Comment:
I don't remember the exact implementation, but the first statement seems inaccurate.

What we do is basically checking if local branch is diverged from the (local) remote-tracking branch, and enable `force` push only when needed. There's a TODO to make `force` push safer by using force-with-lease, but it's not implemented yet.

Reply:
I'm a bit confused and unsure what precisely you mean. I might un-confuse myself, but if you could remind me what you are worried about, it might help. My memory of what happens exactly is hazy; here are a few possibilities of what might be the problem that came to my mind.

Let's say the local branch moved sideways and the remote branch moved somewhere else. I'm pretty sure (or at least I very much hope) `jj git push` would fail in this case. However, if "force push" was enabled, though, because the local branch moved sideways, it would succeed. Do you think that's a possibility? (I might look at the code tomorrow, if I do this will be the first thing I'll think about).

I'm also realizing that what I wrote doesn't take into account the difference between jj's remote-tracking branches and the git repo's remote-tracking branch refs. I hope this difference is not very material, but I could be forgetting something important.

If what I wrote is correct, and the difference between git's and jj's remote tracking branches is immaterial, I'm not sure whether using Git's "force-with-lease" would gain us anything.

Reply:
> Let's say the local branch moved sideways and the remote branch moved somewhere else. I'm pretty sure (or at least I very much hope) `jj git push` would fail in this case. However, if "force push" was enabled, though, because the local branch moved sideways, it would succeed. Do you think that's a possibility? (I might look at the code tomorrow, if I do this will be the first thing I'll think about).

Yes, I'm pretty sure that's a possibility. That's why we want to switch to force-with-lease.

Reply:
I see, thanks. I'll hopefully look into it.

I guess the problem is that we need the Git library's support for the push to both be atomic and to do force-with-lease. Grr.

This might also mean that the answer I gave in https://discord.com/channels/968932220549103686/1226363798483636265 was incorrect.

Reply:
I also found https://github.com/libgit2/libgit2/issues/3947#issuecomment-1915286717 and https://github.com/libgit2/libgit2/issues/4286#issuecomment-527804088.

Reply:
This seems totally doable now 🎉 . See https://github.com/rust-lang/git2-rs/pull/926/files#diff-a9e4250104d90130c09939fc70ad76c0d3db800fc0e42a990b1d73d98009e9cb. I'll try to implement it.

Reply:
What was a lie before is now the truth (after #3522)!

I updated the PR significantly. PTAL.